### PR TITLE
Refactoring/crypto

### PR DESCRIPTION
**Pull request by KiriosK**

### CODE REVIEW

File: primitive/crypto/src/dress/curve/weierstrass/group.rs [hunk 1]
```diff
@@ -77,23 +77,71 @@ macro_rules! affine_group_operation {
...
             type Output = $projective;
 
-            fn mul(self, rhs: &'b <Self as CurveGroup>::Scalar) -> Self::Output {
-                scalar_point(self.into(), &rhs)
+            fn mul(self, rhs: &'a $scalar) -> Self::Output {
+                scalar_point(self.to_extended(), &rhs)
+            }
+        }
+
+        impl<'a> Mul<$scalar> for &'a $affine {
```
Comment (ashWhiteHat) [inline primitive/crypto/src/dress/curve/weierstrass/group.rs:100]: Can we move these to `refops`?

File: primitive/crypto/src/dress/curve/weierstrass/group.rs [hunk 2]
```diff
@@ -256,36 +352,36 @@ macro_rules! curve_arithmetic_extension {
 
         impl<'b> SubAssign<&'b $curve> for $curve {
             fn sub_assign(&mut self, rhs: &'b $curve) {
-                *self = &*self - rhs;
+                *self = (&*self - rhs).into();
```
Comment (ashWhiteHat) [inline primitive/crypto/src/dress/curve/weierstrass/group.rs:355]: I prefer `-=`.

File: primitive/crypto/src/dress/curve/weierstrass/group.rs [hunk 3]
```diff
@@ -256,36 +352,36 @@ macro_rules! curve_arithmetic_extension {
...
         }
 
         impl<'a, 'b> Sub<&'b $curve> for &'a $curve {
-            type Output = $curve;
+            type Output = $projective;
 
-            fn sub(self, rhs: &'b $curve) -> $curve {
-                self - rhs
+            fn sub(self, rhs: &'b $curve) -> $projective {
+                *self - *rhs
```
Comment (ashWhiteHat) [inline primitive/crypto/src/dress/curve/weierstrass/group.rs:363]: Can we use `&` instead?

File: primitive/jubjub/src/curve.rs [hunk 4]
```diff
@@ -102,3 +102,67 @@ impl<'a, 'b> Mul<&'b Fp> for &'a JubjubExtend {
...
+        let _ = ext1 * scalar;
+        let _ = ext1 * &scalar;
+        let _ = &ext1 * scalar;
+        let _ = &ext1 * &scalar;
+        let _ = scalar * ext1;
+        let _ = &scalar * &ext1;
+        let _ = scalar * &ext1;
+        let _ = &scalar * ext1;
+    }
+}
```
Comment (ashWhiteHat) [inline primitive/jubjub/src/curve.rs:168]: It would be nice if we move these to `curve_test`.

**Review by ashWhiteHat — CHANGES_REQUESTED** [review]

File: primitive/bls12_381/src/g1.rs [hunk 5]
```diff
@@ -349,4 +349,59 @@ mod tests {
...
+
+        let _ = ext1 * scalar;
+        let _ = ext1 * &scalar;
+        let _ = &ext1 * scalar;
+        let _ = &ext1 * &scalar;
+        let _ = scalar * ext1;
+        let _ = &scalar * &ext1;
+        let _ = scalar * &ext1;
+        let _ = &scalar * ext1;
+    }
```
[the remark this answers is not in the document]
Comment (ashWhiteHat) [reply primitive/bls12_381/src/g1.rs:406]: Next

File: primitive/crypto/src/dress/curve/weierstrass/group.rs (hunk 3 above)
Comment (ashWhiteHat) [reply primitive/crypto/src/dress/curve/weierstrass/group.rs:363]: not possible

File: primitive/crypto/src/dress/curve/weierstrass/group.rs (hunk 1 above)
Comment (KiriosK) [reply primitive/crypto/src/dress/curve/weierstrass/group.rs]: Done

File: primitive/crypto/src/dress/curve/weierstrass/group.rs (hunk 2 above)
Comment (KiriosK) [reply primitive/crypto/src/dress/curve/weierstrass/group.rs]: Done

File: primitive/crypto/src/behave/curve/basic.rs [hunk 6]
```diff
@@ -22,15 +19,21 @@ pub trait Curve: CurveGroup + ParityCmp + Basic {
...
+pub trait Affine: Curve {
+    fn to_extended(self) -> <Self as CurveGroup>::Extended;
+}
 
 /// extend curve point representation
 /// projective, jacobian and so on
-pub trait CurveExtend:
+pub trait CurveExtended:
     Curve
+    // + RefOps
```
[the remark this answers is not in the document]
Comment (KiriosK) [reply primitive/crypto/src/behave/curve/basic.rs]: Removed

**Pull request merged**